apis/v1alpha1/nginxproxy_types.go

```diff
@@ -32,6 +32,10 @@ type NginxProxySpec struct {
 	//
 	// +optional
 	Telemetry *Telemetry `json:"telemetry,omitempty"`
+	// BaseHTTPConfig specifies the base http context configuration.
+	//
+	// +optional
+	BaseHTTPConfig *BaseHTTPConfig `json:"baseHTTPConfig,omitempty"`
```

Comment:
Hard one to predict...

I feel like this is too broad of a category that could become a dumping ground anyway, so I almost feel that http2 could be its own top level field.

Reply:
as an experiment, we can try to define more features (fields) here in the comment and see how they would all fit together -- https://github.com/nginxinc/nginx-gateway-fabric/blob/main/docs/proposals/nginx-extensions.md#gateway-settings + http3 https://nginx.org/en/docs/quic.html

we can also try to think further ahead and think about configurations for stream module.

Reply:
Yeah I agree that having a generic http context struct probably isn't adding much value. 

That's a good point @pleshakov about http3. There are also other http2 configuration options that we may want to light up later. Should we think about having a struct for just those? `HTTPProtocolConfig` is what I've chosen here (although that's Hypertext Transfer Protocol Protocol Configuration 🙃).

```
type NginxProxySpec struct {
	// Telemetry specifies the OpenTelemetry configuration.
	//
	// +optional
	Telemetry *Telemetry `json:"telemetry,omitempty"`
	// HTTPProtocolConfig specifies the HTTP protocol configuration.
	//
	// +optional
	HTTPProtocolConfig *HTTPProtocolConfig `json:"HTTPProtocolConfig,omitempty"`
    //
    // LogSettings specifies the error and access log settings (http and stream?).
	//
	// +optional
	LogSettings *LogSettings `json:"LogSettings,omitempty"`
    //
    // WorkerSettings specifies the worker settings.
	//
	// +optional
	WorkerSettings *WorkerSettings `json:"WorkerSettings,omitempty"`
}
<...>
// HTTPProtocolConfig specifies the HTTP protocol configuration.
type HTTPProtocolConfig struct {
	// DisableHTTP2 defines if http2 should be disabled for all servers.
	// Default is false, meaning http2 will be enabled for all servers.
	//
	// +optional
	DisableHTTP2 bool `json:"disableHTTP2"`
	// other http2/ http3/ QUIC settings as needed
}
```

Reply:
Logging definitely fits under a struct. `proxy_protocol` feels standalone. `http3` feels like it could be its own struct. Which leaves `http2` by itself. `HTTPProtocolConfig` is definitely a mouthful, and based on our current predictions on possible fields to go in this CRD, I don't know if it's a necessary struct to have.

I don't think we want any sort of `http` and `stream` structure in here because we should try our best to encapsulate nginx functionality so users don't need to know about those separate contexts.

Reply:
There are four additional http2 directives that we could potentially light up: 
- `http2_body_preread_size` (contexts: http, server)
- `http2_chunk_size` (contexts: http, server, location)
- `http2_max_concurrent_streams` (contexts: http, server)
- `http2_recv_buffer_size` (contexts: http, server)

I think we need to decide whether it makes sense to expose these directives in the NginxProxy CRD. If it does, we should add an `http2` struct that can grow to include these fields. If not, I suggest going with a top-level field. It doesn't seem to fit with any of the other planned features, like logging or worker settings.

For what it's worth, NIC doesn't expose the http2 directives listed above. It may be that their use cases aren't strong enough to warrant a first-class API. 

I would also be fine going with a top-level field for now, and if we decide to add these http2 directives in the future, moving that field under a struct and deprecating the top-level field.

Reply:
> I don't think we want any sort of http and stream structure in here because we should try our best to encapsulate nginx functionality so users don't need to know about those separate contexts.

log format will be different for http and stream 
http2 or http3 setting doesn't apply to stream.

here is how potentially the resource with stream and http top level fields can look.

```
http:
   http2: ...
   http3: ...
   logging:
     accessLogFormat: ...
stream:
   logging:
     accessLogFormat: ...
worker: ....
logging:
 error: ... # error log is common for both http and stream
```

Reply:
I still don't know how I feel about `http` and `stream` blocks though...unless we're assuming non-nginx users are going to know what those contexts mean? Per my last comment, encapsulating nginx functionality is one of our desires with these CRDs.

But if there isn't another logical way to separate configs, then I'm not sure

Reply:
RE logging: an alternative:
```
  logging:
     accessLogFormat: ...
worker: ....
logging:

```

Reply:
RE logging: an alternative:
```
logging:
    accessLogFormat:
        http: ...
        stream: ...
    errorLog: ...
worker: ....
```

Reply:
For logging, I was thinking almost exactly along the same lines as you @kate-osborn

It seems the feeling for now is to go with top-level field for disableHTTP2, and optionally move it later if it makes sense. So,

```
type NginxProxySpec struct {
  // Telemetry specifies the OpenTelemetry configuration.
  //
  // +optional
  Telemetry *Telemetry `json:"telemetry,omitempty"`
  // DisableHTTP2 defines if http2 should be disabled for all servers.
  // Default is false, meaning http2 will be enabled for all servers.
  //
  // +optional
  DisableHTTP2 bool `json:"disableHTTP2"`
}
<...>
```